Adaptions for Planning with Submaps

gasserl opened the pull request:
**Changes:**
- Nodes for TSDF mapping server and ESDF mapping server
- Templated submap server class to allow for ESDF mapping
- Interfacing class for submap collection class used for the different submap classes
- Publishing of submaps and pose updates to be used in other nodes, including serialization functions and a new package for messages
- Multithreading of postprocessing such as ESDF generation and mesh visualization to maintain mapping speed
- Improved submap visualizer, now able to visualize any submap and not only the active submap
- Protobuf for saving maps for all submap classes
- Various adaptions to allow for use in the planner, such as accessor functions

**Comments:**
Git does not seem to like renaming, the changes in the submap server can be seen by comparing the [new templated file](https://github.com/ethz-asl/cblox/blob/gasserl/feature/planning_infrastructure/cblox_ros/include/cblox_ros/submap_server_inl.h) with [the old file](https://github.com/ethz-asl/cblox/blob/master/cblox_ros/src/tsdf_submap_server.cc).

gasserl committed:
timings on sending submaps and global map

gasserl committed:
timings on sending submaps and global map part 2

gasserl committed:
pose message, improved timing files

gasserl committed:
finishing touches

gasserl committed:
recording times for submaps

gasserl committed:
derived class TsdfEsdfSubmapServer

gasserl committed:
bugfix

gasserl committed:
version direct inheritance

gasserl committed:
version proper inheritance

gasserl committed:
version template

gasserl committed:
simple constructor for submap (unused)

gasserl committed:
properer definition of template class (not working yet)

gasserl committed:
additional function to create maps based on existing ones

gasserl committed:
cleanup template experiment

alexmillane committed:
Templated the collection and added a simple interface.

gasserl committed:
Merge pull request #12 from ethz-asl/feature/templated_collection

Templated the collection and added a simple interface.

gasserl committed:
templated SaveTsdfSubmapCollection

gasserl committed:
templated submap message conversions

gasserl committed:
backup before converting to only templated submap server

gasserl committed:
move to template server successful

gasserl committed:
planning submap

gasserl committed:
calculate submap bounds

gasserl committed:
proper map bounds calculation

gasserl committed:
included skeleton graph to submap

gasserl committed:
no delay when loading map

gasserl committed:
regeneration of skeleton possible

gasserl committed:
prm generator, esdf generation flags

gasserl committed:
minor changes

gasserl committed:
/.idea in .gitignore

gasserl committed:
Merge branch 'master' into gasserl/feature/planning_infrastructure

gasserl committed:
minor cleanup

gasserl committed:
beautification

gasserl committed:
gitignore include .idea files

gasserl committed:
resolve merge with master

gasserl committed:
resolve merge conflict (builds)

gasserl committed:
incremental prm growth possible

gasserl committed:
Merge branch 'master' into gasserl/feature/planning_infrastructure

gasserl committed:
Merge branch 'gasserl/feature/planning_infrastructure' of github.com:ethz-asl/c-blox into gasserl/feature/planning_infrastructure

gasserl committed:
debug messages

gasserl committed:
extracted planning dependency

gasserl committed:
pull merge

gasserl committed:
less build warnings

gasserl committed:
ESDF protobufs

gasserl committed:
bugfix

gasserl committed:
less build warnings

gasserl committed:
templating attempt

gasserl committed:
protobuf builds

gasserl committed:
more protobuf

gasserl committed:
renames and cleanups

gasserl committed:
msgs according to voxgraph

gasserl committed:
tsdf/esdf flag in msg, pause during publishing of entire map

gasserl committed:
corrected indents

gasserl committed:
multithreading finishing submap

gasserl committed:
multithread esdf generation, esdf slice, reorg

gasserl committed:
launch file

gasserl committed:
correction reorg

gasserl committed:
lock preparations

gasserl committed:
pose updates

gasserl committed:
mesh from MarkerArray, verbose attribute

gasserl committed:
submap service

gasserl committed:
ESDF mutex, proper active submap initialization

gasserl committed:
updated dependencies

gasserl committed:
submap service

gasserl committed:
threads and locks

gasserl committed:
generalized submap visualizer

gasserl committed:
pose updates

gasserl committed:
generally obtaining rosparams

gasserl committed:
general submap mesh visualizer

gasserl committed:
voxgraph compatible pose updates

gasserl committed:
visualization only if topic has subscribers

gasserl committed:
better visualization

gasserl committed:
visualization (de-roof, pastellize, color cycle length)

gasserl committed:
resolved build error due to old ros msg

gasserl committed:
shuffle I

gasserl committed:
shuffle II

gasserl committed:
shuffle III

gasserl committed:
minimize changes and cleanup

gasserl committed:
cleanup cblox from ros dependencies

gasserl committed:
tsdf server node and correctly linking template specialization

gasserl committed:
transparent map mesh

gasserl committed:
service call to publish submap poses

gasserl committed:
temp

gasserl committed:
temp 2

gasserl committed:
voxblox protobuf update

alexmillane commented on cblox/CMakeLists.txt:
```diff
@@ -27,7 +27,8 @@ if (protobuf_catkin_FOUND)
 
-    PROTOBUF_CATKIN_GENERATE_CPP(PROTO_SRCS PROTO_HDRS ${PROTO_DEFNS})
+    set(BASE_PATH "proto/cblox")
+    PROTOBUF_CATKIN_GENERATE_CPP2(${BASE_PATH} PROTO_SRCS PROTO_HDRS ${PROTO_DEFNS})
```
We changed the way we're compiling the protobuf stuff in master, which is conflicting here. But let's sort that out once we're ready to merge.

alexmillane commented on cblox/include/cblox/core/submap_collection.h:
```diff
@@ -6,29 +6,57 @@
 #include <vector>
 
-#include "./TsdfSubmapCollection.pb.h"
+#include "./SubmapCollection.pb.h"
```
All of the protobuf includes are less hideous now. I guess this is where the conflicts on these files are coming from.

alexmillane commented on cblox/include/cblox/core/submap_collection.h:
```diff
@@ -6,29 +6,57 @@
+  // NOTE(alexmillane): I'm moving methods over only as I need them. There's no
+  // design intent here in leaving some out. There is only the intent to be
+  // lazy.
+  virtual const Transformation &getActiveSubmapPose() const = 0;
```
The reference symbol alignment changes throughout this project. The intent is left, here (and elsewhere) it's right.

alexmillane commented on cblox/include/cblox/core/submap_collection.h:
```diff
@@ -6,29 +6,57 @@
   // Gets a vector of the linked IDs
   std::vector<SubmapID> getIDs() const;
-  bool exists(const SubmapID submap_id) const;
+  bool exists(const SubmapID& submap_id) const;
```
SubmapID is a typedef of an integral type and therefore is passed by const value, not reference.

alexmillane commented on cblox/include/cblox/core/submap_collection_inl.h:
```diff
@@ -38,7 +38,7 @@ std::vector<SubmapID> SubmapCollection<SubmapType>::getIDs() const {
 
 template <typename SubmapType>
-bool SubmapCollection<SubmapType>::exists(const SubmapID submap_id) const {
+bool SubmapCollection<SubmapType>::exists(const SubmapID& submap_id) const {
```
by value

alexmillane commented on cblox/include/cblox/core/submap_collection_inl.h:
```diff
@@ -162,7 +171,7 @@ const Transformation& SubmapCollection<SubmapType>::getActiveSubmapPose()
 }
 template <typename SubmapType>
-const SubmapID SubmapCollection<SubmapType>::getActiveSubmapID() const {
+const SubmapID& SubmapCollection<SubmapType>::getActiveSubmapID() const {
```
value

alexmillane commented on cblox/include/cblox/core/submap_collection_inl.h:
```diff
@@ -285,20 +294,22 @@ bool SubmapCollection<SubmapType>::saveToFile(
     // Saving the submap
-    (id_submap_pair.second)->saveToStream(&outfile);
+    bool success = (id_submap_pair.second)->saveToStream(&outfile);
+    if (success) {LOG(INFO) << "Saving successful";}
+    else {LOG(WARNING) << "Saving unsuccessful";}
```
This formatting is disallowed by google style guide:

https://google.github.io/styleguide/cppguide.html#Conditionals

I'm surprised your formatter didn't pick this up... Maybe you missed this file?

Probably worth one of us running the formatter on the entire project before merging this.

alexmillane commented on cblox/include/cblox/core/submap_collection_inl.h:
```diff
@@ -340,6 +351,9 @@ void SubmapCollection<SubmapType>::fuseSubmapPair(
     const Transformation T_S1_S2 = T_G_S1.inverse() * T_G_S2;
     // Merging the submap layers
+
+    std::unique_lock<std::mutex> lock_submap_1(submap_ptr_1->submap_mutex);
+    std::unique_lock<std::mutex> lock_submap_2(submap_ptr_2->submap_mutex);
```
From what I can see, <submap_mutex> is never defined anywhere... Looking deeper I realized this function is never actually called anywhere and therefore never instantiated. That's why it compiles despite using a non-existing member.

Despite the fact it's not used, I would leave it in. For now let's just remove the locks, and maybe add a note somewhere that the function is not threadsafe.

ntonci replied:
The other option is to add mutex to the submap class. I guess that would be useful if one would want to speed up the getProjectedMap function by using this method and parallelizing it. Wdyt?

alexmillane replied:
I agree having a mutex here makes sense somehow. I see that you've now added it. The idea seems reasonable to me.

The issue as it is right now is that: because submap_mutex submap mutex is private, it cannot be accessed directly from another class like this. There must be a pattern for this kind of problem but I'm not familiar enough with multithreaded stuff to know it. I could think of a member function returning a unique_lock?

Any ideas @victorreijgwart

alexmillane replied:
Having taken a closer look at this I would prefer to remove the locks. Currently, the only function that actually uses them is never instantiated, hence why coding errors have slipped through a couple of times. In light of that, I would prefer to add the locks when they're actually needed and can be tested.

victorreijgwart replied:
If we intend to keep these locks, it'd be important to carefully think them through. Given that they easily create (very nasty) bugs. So unless we really need them, I'd also be in favor of removing them for now (instead of leaving booby traps in the code). We can then carefully reintroduce them later.

alexmillane commented on cblox/include/cblox/core/tsdf_submap.h:
```diff
@@ -50,6 +50,18 @@ class TsdfSubmap {
+  void stopMappingTime(int32_t time) {
+    mapping_interval_.second = time;
+  }
+  // Access mapping interval
+  const std::pair<int32_t, int32_t>& getMappingInterval() const {
+    return mapping_interval_;
+  }
```
Nice addition!

alexmillane reviewed (commented):
@gasserl 

I'm going to slowly look at this in pieces over the next few days. Will let you know when I'm finished. Great work!

alexmillane commented on cblox/include/cblox/io/tsdf_submap_io_inl.h:
```diff
@@ -8,65 +8,83 @@
 namespace io {
 
-template <typename SubmapType>
+// we assume most SubmapTypes to have an ESDF map
```
I feel that this is a bit back-to-front. Voxblox _requires_ a TSDF layer but not an ESDF layer is optional. For example in recently (darpa?) people have used occupancy layers, and in [voxblox++](https://github.com/ethz-asl/voxblox-plusplus) they're using semantic layers, both in addition to TSDF but not in addition to an ESDF.

ntonci replied:
I guess since now both layers are always in the proto message, it makes sense that we just try to read everything (it shouldn't matter if it has only tsdf or only esdf or nothing, the number of blocks in the proto header will just be 0 then)

ntonci replied:
Ok now I see what you meant :)

alexmillane commented on cblox/include/cblox/io/tsdf_submap_io_inl.h:
```diff
@@ -8,65 +8,83 @@
+    return false;
+  }
+
+  // Getting the esdf blocks for this submap (the tsdf layer)
```
Update the comment to "the esdf layer"

alexmillane commented on cblox/src/core/tsdf_esdf_submap.cpp:
```diff
@@ -9,7 +10,41 @@ void TsdfEsdfSubmap::generateEsdf() {
   esdf_integrator.updateFromTsdfLayerBatch();
 }
 
+void TsdfEsdfSubmap::setTsdfMap(const voxblox::Layer<TsdfVoxel>& tsdf_layer) {
+  tsdf_map_.reset(new voxblox::TsdfMap(tsdf_layer));
+}
```
Where do we use this function? Are there times where we actually switch a submap's layer after it's creation.

The issue with this is that if we do need this, we open the door to inconsistencies in the class. For example the esdf and the tsdf being mismatched in some way.

ntonci replied:
Yeah I agree, this seems a bit dangerous, I guess I will just remove it for now and in case there is a use-case for it we can bring it back.

alexmillane replied:
Cool!

alexmillane commented on cblox/src/core/tsdf_esdf_submap.cpp:
```diff
@@ -9,7 +10,41 @@ void TsdfEsdfSubmap::generateEsdf() {
                                           esdf_map_->getEsdfLayerPtr());
   // Generate the ESDF
   LOG(INFO) << "Generating ESDF from TSDF for submap with ID: " << submap_id_;
+  esdf_integrator.setFullEuclidean(false);
```
This is a parameter in the `esdf_integrator_config_`. Taking it from the params has the advantage that it is configurable. Is there any reason that we need this had coded here?

alexmillane commented on cblox/src/core/tsdf_esdf_submap.cpp:
```diff
@@ -9,7 +10,41 @@ void TsdfEsdfSubmap::generateEsdf() {
+  // add ESDF info
+  size_t num_blocks = esdf_map_->getEsdfLayer().getNumberOfAllocatedBlocks();
+  proto->set_num_esdf_blocks(num_blocks);
+}
```
space between method definitions.

alexmillane commented on cblox_msgs/msg/MapLayer.msg:
```diff
@@ -0,0 +1,15 @@
+# Message header
+Header header
+
+# Map header encapsulating the map's main properties
+cblox_msgs/MapHeader map_header
+
+# Layer type
+# 0 - TSDF
+# 1 - ESDF
+uint8 type
+
+# Layer containing the map's TSDF or ESDF
+# NOTE: It is expressed in the frame specified by header.frame_id
+voxblox_msgs/Layer tsdf_layer
+voxblox_msgs/Layer esdf_layer
```
Great! Thanks for this!

alexmillane commented on cblox_msgs/package.xml:
```diff
@@ -0,0 +1,26 @@
+  <version>0.0.0</version>
+  <description>Message definitions for Cblox ROS interface.</description>
+
+  <author email="helen.oleynikova@mavt.ethz.ch">Helen Oleynikova</author>
+  <author email="marius.fehr@mavt.ethz.ch">Marius Fehr</author>
+  <author email="gasserl@student.ethz.ch">Laura Gasser</author>
+
+  <maintainer email="helen.oleynikova@mavt.ethz.ch">Helen Oleynikova</maintainer>
+  <maintainer email="marius.fehr@mavt.ethz.ch">Marius Fehr</maintainer>
```
I guess let's change this to you and I? Not that anyone has probably ever even looked at these definitions.

alexmillane commented on cblox_ros/include/cblox_ros/active_submap_visualizer.h:
```diff
@@ -18,29 +19,32 @@ using voxblox::MeshIntegrator;
   void getDisplayMesh(visualization_msgs::Marker* marker_ptr);
   MeshLayer::Ptr getDisplayMeshLayer();
 
+  void setVerbose(const bool& verbose) {verbose_ = verbose;}
+  void setOpacity(const float& opacity) {opacity_ = opacity;}
```
format

alexmillane reviewed (commented):
Bit more progress.

alexmillane commented on cblox_ros/include/cblox_ros/ros_params.h:
```diff
@@ -31,6 +35,20 @@ inline voxblox::TsdfIntegratorType getTsdfIntegratorTypeFromRosParam(
+inline typename TsdfSubmap::Config getSubmapConfigFromRosParam<TsdfSubmap>(
+    const ros::NodeHandle& nh_private) {
+  return voxblox::getTsdfMapConfigFromRosParam(nh_private);
+}
```
I would again here invert the esdf/tsdf specialization logic.

alexmillane commented on cblox_ros/include/cblox_ros/submap_conversions_inl.h:
```diff
@@ -0,0 +1,150 @@
+  ros::Time timestamp = ros::Time::now();
+  // fill in headers
+  msg->header = generateHeaderMsg<SubmapType>(submap_ptr, timestamp);
+  msg->map_headers[submap_ptr->getID()] = generateSubmapHeaderMsg<SubmapType>(submap_ptr);
```
There's going to be an issue here if there's non-consecutive submap ids. I'm guessing we've implicitly assumed this elsewhere in the code. Leave this for now, because I suspect it's going to be difficult to change, however, maybe make a note that we're assuming the submap IDs are equal to their index in storage.

alexmillane commented on cblox_ros/include/cblox_ros/submap_conversions_inl.h:
```diff
@@ -0,0 +1,150 @@
+}
+
+template<typename SubmapType>
+SubmapID deserializeMsgToPose(const cblox_msgs::MapPoseUpdate* msg,
```
const ref to the msg?

alexmillane commented on cblox_ros/include/cblox_ros/submap_conversions_inl.h:
```diff
@@ -0,0 +1,150 @@
+    submap_id = *std::max(submap_ids.begin(), submap_ids.end());
+  }
+  return submap_id;
+}
```
As I read it, this function does a number of things. It deserializes a pose update message as advertised. But it also updates the submap collection. I would say this should be split into two functions. The deserialize function returning a list of submap_id, transform pairs and another function updating the submap collection.

alexmillane replied:
I see now that you were copying the interface from voxblox `deserializeMsgToLayer`. I think in this case, because updating the collection is non-trivial, it makes sense to separate the deserialization from the update. What's your opinion?

alexmillane commented on cblox_ros/include/cblox_ros/submap_conversions_inl.h:
```diff
@@ -0,0 +1,150 @@
+  msg->map_header = generateSubmapHeaderMsg<SubmapType>(submap_ptr);
+
+  // set type to TSDF
+  msg->type = 0;
```
I would add an enum or constexpr somewhere the header which gives this value a name.

alexmillane replied:
Something like
```
enum class MapLayerTypes : uint8_t {
  kTsdf = 0u,
  kEsdf = 1u
};
```

alexmillane commented on cblox_ros/include/cblox_ros/submap_conversions_inl.h:
```diff
@@ -0,0 +1,150 @@
+  return submap_collection_ptr->getSubmapPtr(submap_id);
+}
+
+// Note: Specialized for TsdfEsdfSubmap
```
Is this comment incorrect?

alexmillane commented on cblox_ros/include/cblox_ros/submap_server.h:
```diff
@@ -1,24 +1,29 @@
 #include <sensor_msgs/PointCloud2.h>
 #include <std_srvs/Empty.h>
+#include <cblox_msgs/SubmapSrvRequest.h>
+#include <cblox_msgs/SubmapSrvResponse.h>
```
Move these down to join the other `cblox_msgs` includes.

alexmillane commented on cblox_ros/include/cblox_ros/submap_server.h:
```diff
@@ -61,12 +67,20 @@ class TsdfSubmapServer {
+  inline typename SubmapCollection<SubmapType>::Ptr
+      getSubmapCollectionPtr();
+  inline typename SubmapCollection<SubmapType>::ConstPtr
+      getSubmapCollectionPtr() const;
```
I think to inline things you might have to move the definition here, into the header.

alexmillane commented on cblox_ros/include/cblox_ros/submap_server_inl.h:
```diff
@@ -0,0 +1,802 @@
+  trajectory_visualizer_ptr_.reset(new TrajectoryVisualizer);
+}
+
+template<>
+SubmapCollection<TsdfSubmap>::Ptr
+    SubmapServer<TsdfSubmap>::getSubmapCollectionPtr();
+template<>
+SubmapCollection<TsdfSubmap>::ConstPtr
+SubmapServer<TsdfSubmap>::getSubmapCollectionPtr() const;
+template<>
+SubmapCollection<TsdfEsdfSubmap>::Ptr
+    SubmapServer<TsdfEsdfSubmap>::getSubmapCollectionPtr();
+template<>
+SubmapCollection<TsdfEsdfSubmap>::ConstPtr
+SubmapServer<TsdfEsdfSubmap>::getSubmapCollectionPtr() const;
```
Do these require specialization?

alexmillane commented on cblox_ros/include/cblox_ros/submap_server_inl.h:
```diff
@@ -0,0 +1,802 @@
+  pointcloud_sub_ = nh_.subscribe("pointcloud", pointcloud_queue_size,
+                                  &SubmapServer<SubmapType>::pointcloudCallback, this);
+  // Subscribing to new submaps and submap pose updates
+  int submap_queue_size = 20;
```
`constexpr int kSubmapQueueSize = 20;`

alexmillane commented on cblox_ros/include/cblox_ros/submap_server_inl.h:
```diff
@@ -0,0 +1,802 @@
+  }
+}
+
+// Note: Specialized for TsdfSubmap
+template<typename SubmapType>
+void SubmapServer<SubmapType>::finishSubmap(const SubmapID& submap_id) {
+  if (submap_collection_ptr_->exists(submap_id)) {
+    typename SubmapType::Ptr submap_ptr =
+        submap_collection_ptr_->getSubmapPtr(submap_id);
+    // stopping the mapping interval
+    submap_ptr->stopMappingTime(ros::Time::now().toSec());
+    // generating ESDF map
+    submap_ptr->generateEsdf();
+    // publishing the old submap
+    publishSubmap(submap_id);
+    ROS_INFO("[CbloxServer] Finished submap %d", submap_id);
+  }
+}
+template<>
+void SubmapServer<TsdfSubmap>::finishSubmap(const SubmapID& submap_id);
```
Rather than a template specialization, I would use a virtual function in the submap subclasses (lol). In addition to being more interpretable, this falls nicely inline with what we've done in voxgraph:

https://github.com/ethz-asl/voxgraph/blob/master/voxgraph/include/voxgraph/frontend/submap_collection/voxgraph_submap.h#L55

(although we should make that function virtual and override ;)

alexmillane commented on cblox_ros/include/cblox_ros/submap_server_inl.h:
```diff
@@ -0,0 +1,802 @@
+  return true;
+}
+
+template<typename SubmapType>
+bool SubmapServer<SubmapType>::publishActiveSubmap() {
+  if (submap_collection_ptr_->empty()) {
+    ROS_WARN("[CbloxPlanner] Active submap does not exist!");
+    return false;
+  }
+  SubmapID submap_id = submap_collection_ptr_->getActiveSubmapID();
+  submap_collection_ptr_->getSubmapPtr(submap_id)->generateEsdf();
+  publishSubmap(submap_id);
+  return true;
+}
+template<>
+bool SubmapServer<TsdfSubmap>::publishActiveSubmap();
```
I guess we have two implementations here because of the call to `generateEsdf()`?

because the two specializations share quite a lot of code, I would move keep this function unspecified and make another function, maybe `prepareSubmapForPublish(SubmapType submap)` and put the specialized code in there.

ntonci replied:
I also made it virtual same as `finishSubmap`, although it is doing the same (generate esdf), I created a new method `prepareForPublish` since it might differ (from `finishSubmap`) in the derived class

alexmillane commented on cblox_ros/include/cblox_ros/submap_server_inl.h:
```diff
@@ -0,0 +1,802 @@
+  for (const SubmapID submap_id : submap_collection_ptr_->getIDs()) {
+    publishSubmap(submap_id);
+    // delay to allow for processing of sent maps
+    ros::Duration sleepy(0.1);
```
lolz. noice.

alexmillane reviewed (commented):
Almost done

ntonci committed:
Merge branch 'master' into gasserl/feature/planning_infrastructure

ntonci committed:
Style fix

ntonci committed:
Fix proto include and merge conflict leftovers

ntonci committed:
Addressed PR comments

- Remove ref rom SubmapID
- Create a new id-pose map for setSubmapPoses method
- Add mutex to each submap (submap_mutex_)
- Added finishSubmap and prepareForPublish methods to submaps
- When loading submap from stream assume it only has a tsdf layer
- Update package.xml authors and maintainers
- Add MapLayerTypes enum class for message filling
- Remove specialized templates for getSubmapCollectionPtr method
- Split deserializeMsgToPose method

ntonci commented:
I think I addressed all the comments from @alexmillane. There are couple of questions/comments left:
- Do we keep the mutex in fuseSubmapPair?
- setTsdfMap was removed, is this fine?
- I guess you need to add override for finishSubmap in Voxblox and you need an implementation of prepareForPublish (unless you are happy with the one from TsdfEsdfSubmap)
- If saving function is a method in submap type why is loading not? This would again make it a bit simpler since you would not need template specializations.

In general, would be cool if someone could test this :)

alexmillane committed:
Removed concurrency gaurds on submap fusion.

alexmillane committed:
Add override keyword.

alexmillane commented:
Thanks for that @ntonci! Looks good to me! Comments:

> Do we keep the mutex in fuseSubmapPair?

I was going back and forth on this. Obviously they _should_ be in there. With that said, I feel that it's best we implement them when someone actually wants them. Mainly just because in that case, someone will test them when they implement them.

> setTsdfMap was removed, is this fine?

I think the way it was, was dangerous. I think we remove this (as you did) and implement it safely if we need it. 

> I guess you need to add override for finishSubmap in Voxblox and you need an implementation of prepareForPublish (unless you are happy with the one from TsdfEsdfSubmap)

Added `virtual ... override` in voxgraph (see: ethz-asl/voxgraph#21)

> If saving function is a method in submap type why is loading not? This would again make it a bit simpler since you would not need template specializations.

Hmm agreed, this is getting quite messy with a lot of repeated code. I'll have a go now, and make a PR onto this one when I get something.

alexmillane committed:
WIP. Checking something back on master.

alexmillane committed:
First compiling implementation of submap loading in subclasses.

alexmillane committed:
Updating the io file naming to better reflect the new design.

alexmillane commented:
Response to comment above is in #29

alexmillane committed:
Fixed the small PR comments.

alexmillane committed:
Moved loading implementation into the collection class.

alexmillane committed:
Merge pull request #29 from ethz-asl/feature/loading_in_submap_classes

Feature/loading in submap classes

alexmillane committed:
Removed now empty loading implementation file.

alexmillane committed:
Updated to return by value

ntonci commented on cblox/include/cblox/io/submap_io.h:
```diff
@@ -14,15 +14,18 @@ namespace io {
-                          const std::string& file_path);
+                          const std::string& file_path) {
+  return submap_collection.saveToFile(file_path);
+}
```
I guess you can make them inline now

ntonci commented:
I see that there are three more functions in SubmapCollection:
`getProjectedMap`, `duplicateSubmap` and `fuseSubmapPair`
that are just using tsdf layer, should this somehow be handled by submaps themselves such that we don't need specialized templates in submapCollection again.
I'm just mentioning this for the sake of completeness, we can also open it as an issue and leave for the next PR.

ntonci commented:
One more thing, I guess it would be nice to also have a base Submap class such that the users know what should be implemented in the derived classes. 
This is also more towards nice to have future PR things :)

victorreijgwart commented:
> One more thing, I guess it would be nice to also have a base Submap class such that the users know what should be implemented in the derived classes.
> This is also more towards nice to have future PR things :)

+1 This would be really nice.

alexmillane committed:
Merge pull request #30 from ethz-asl/feature/return_by_value_load_from_stream

Updated to return by value

victorreijgwart commented on cblox/include/cblox/core/tsdf_submap.h:
```diff
@@ -41,15 +41,23 @@ class TsdfSubmap {
   }
 
+  // Set interval in which submap was actively mapping
+  void startMappingTime(int32_t time) { mapping_interval_.first = time; }
```
Would you usually use this as seconds since epoch? In some cases I'd expect something more accurate to be quite useful. How about using an `int64_t`? Compared to the size of the rest of the submap, it wouldn't be noticeable in terms of storage. And being able to store full system clock timestamps (nanosecond precision) could be quite useful :)

alexmillane commented:
> One more thing, I guess it would be nice to also have a base Submap class such that the users know what should be implemented in the derived classes.
> This is also more towards nice to have future PR things :)

Yeah I totally agree here.

victorreijgwart commented on cblox_msgs/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(cblox_msgs)
+
+find_package(catkin REQUIRED COMPONENTS
+    geometry_msgs
+    voxblox_msgs
+    message_generation
+)
+include_directories(include ${catkin_INCLUDE_DIRS})
+
+add_message_files(
+    FILES
+    MapHeader.msg
+    MapLayer.msg
+    MapPoseEstimate.msg
+    MapPoseUpdate.msg
+)
+
+add_service_files(
+    FILES
+    SubmapSrv.srv
+)
+
+generate_messages(
+    DEPENDENCIES
+    geometry_msgs
+    voxblox_msgs
+)
+
+catkin_package(
+    CATKIN_DEPENDS message_runtime
+)
+
+##########
+# EXPORT #
+##########
+
+#cs_install()
+#cs_export()
```
Is there a reason to not just use `catkin_simple`?

alexmillane replied:
done!

victorreijgwart commented on cblox_msgs/msg/MapPoseUpdate.msg:
```diff
@@ -0,0 +1,5 @@
+# Message header
+Header header
+
+# Map header encapsulating the map's main properties
+cblox_msgs/MapHeader[] map_headers
```
From the current name of this message it's not immediately clear that it contains pose updates for an array of submaps, instead of just a single one. What about calling this message `MapHeaderList` (clearest as to what it contains), `MapPoseList` (clearer in terms of its main use-case), or perhaps simply `MapPoseUpdates`?

alexmillane replied:
I think the pluralization, `MapPoseUpdates` is the best fit here. I will change it to that.

victorreijgwart commented on cblox_ros/src/submap_server.cc:
```diff
@@ -0,0 +1,89 @@
+  color_msg.a = 1.0;
+
+  float max_dist = 0.6;
+  nh_private_.param("truncation_distance", max_dist, max_dist);
```
Like this, the ROS master gets queried every time `SubmapServer<TsdfSubmap>::visualizeSlice(...)` is executed.
Do you intend to change this param at runtime? If yes, it might be worth using [cached params](http://wiki.ros.org/roscpp/Overview/Parameter%20Server#Cached_Parameters). Otherwise it might be nicer to make `max_dist` a class variable and move its loading to [getParametersFromRos()](https://github.com/ethz-asl/cblox/blob/85c44b4351a91745801c7bd4c15245d3e8103b96/cblox_ros/include/cblox_ros/submap_server.h#L110).

alexmillane replied:
Done!

alexmillane committed:
Switched cblox_msgs to catkin_simple.

alexmillane committed:
Changed a message name and added truncation distance as server member.

alexmillane commented:
@ntonci @victorreijgwart I'm going to merge this now to bring master in line for the voxgraph release. There are still some areas of messiness frankly, but let's solve them as we go.
